### PR TITLE
Phase-1 interface to store L1 CSC TPG params in DB

### PR DESCRIPTION
**Pull request by dildick**
The Phase-1 upgrade of CSC TPG includes new ALCT, CLCT and TMB algorithms for ME1/1. Setters and getters were added to update the params in the DB. Clarification: CSCs other than ME1/1 will not be upgraded for Phase-1. ME2/1, ME3/1 and ME4/1 will also be upgraded for Phase-2 (not included in this pull request).

### CODE REVIEW

**Comment by davidlange6** [comment]
please test

**Comment by ggovi** [comment]
@dildick 
you are adding data members to the class CSCDBL1TPParameters
This class has been used in the past for populating tags included in production GTs ( 71 and 74 series ).  
Altering the 'shape' would potentially generates mismatches between reader version and stored payload version, and it should be avoided.
I would recommend to create a new CondFormat class having the new attributes.

**Comment by dildick** [comment]
@ggovi You want me to leave `CSCDBL1TPParameters` untouched, and instead make a class 
`CSCDBL1TPPhase1Parameters` for the phase-1 CSC upgrade? In the far future, the CSC OTMB will be upgraded again and GEM parameters will enter the game. Should we then construct a class `CSCDBL1TPPhase2Parameters`?

**Comment by dildick** [comment]
@slava77 what is your take on this? Should we extend the current `CSCDBL1TPParameters` or make a version of phase-1 upgrade `CSCDBL1TPPhase1Parameters`?

**Comment by davidlange6** [comment]
hi @dildick - just to clarify- you are not allowed to schema evolve conditions database objects, eg, CondFormats (per policy of alca/db)

**Comment by dildick** [comment]
ok. I'll rework this PR then.

**Comment by dildick** [comment]
@ggovi please let me know if this is better

**Comment by arunhep** [comment]
please test

**Comment by dildick** [comment]
@ptcox So here, `Phase1` upgrade refers to the high-lumi algos for ME1/1 (without GEM). If `Phase1` (and `Phase2`) could cause confusion, please let me know if you propose an alternative naming scheme for the following scenarios:
- upgrade ME1/1 
- upgrade ME1/1 with GEM
- upgrade ME2/1 with GEM
- upgrade ME3/1, ME4/1

**Comment by ptcox** [comment]
Re. the naming.I suggest avoiding the use of Phase 1 and 2 terminology in combination with 'upgrade' when they are different from the long-standing existing CMS meaning. You could use TP upgrade step 1-4, or stage 1-4, or 'level 1-4', or anything that avoids existing terms. Maybe Slava and Andrea have some view on this. I'd go with 'stage N'.

**Comment by slava77** [comment]
On 9/8/17 9:13 AM, ptcox wrote:
> Re. the naming.I suggest avoiding the use of Phase 1 and 2 terminology
> in combination with 'upgrade' when they are different from the
> long-standing existing CMS meaning. You could use TP upgrade step 1-4,
> or stage 1-4, or 'level 1-4', or anything that avoids existing terms.
> Maybe Slava and Andrea have some view on this. I'd go with 'stage N'.
> 

[It's not reco code. So, the DB experts have the final say]

IIUC, most of the evolution is limited by lack of schema evolution
capability for data members of the DB types.
I'd changing the existing class names to Extended and put a flexible
enough data member setup so that it can handle the possible additions so
that streaming can be supported "forever" (save a vector of values
aligned with a vector of names or enum-ints that describe what the
values are).
The access methods could evolve with a version or a year affixed to the
name if it's needed as an interface.

**Comment by dildick** [comment]
HI @slava77, is there an example in cmssw that closely matches your description that I can look at?

**Comment by dildick** [comment]
@slava77 please let me know if you can point me to a clear example

**Comment by slava77** [comment]
On 9/12/17 4:42 PM, Sven Dildick wrote:
> @slava77 <https://github.com/slava77> please let me know if you can
> point me to a clear example

I'm not the best reference on this subject, sorry.
My earlier comments were rather generic.

**Comment by dildick** [comment]
@arunhep @lpernie do you have any suggestions?

**Comment by dildick** [comment]
@slava77 I'm sorry, but without a decent example I can't really continue

**Comment by dildick** [comment]
@ggovi @slava77 What if I collect all private data members in a struct. Would that be flexible enough? I can't seem to find an example in CondFormats, so I could be mistaken

**Comment by slava77** [comment]
On 11/28/17 2:46 PM, Sven Dildick wrote:
> @ggovi <https://github.com/ggovi> @slava77 <https://github.com/slava77>
> What if I collect all private data members in a struct. Would that be
> flexible enough? I can't seem to find an example in CondFormats, so I
> could be mistaken

IIRC, the reason to introduce a the generic container was to effectively
implement schema evolution, which is not supported at cond object
streaming level.
If you think that a fixed set of variables in a struct can support all
possible cases, then the struct should be fine.

**Comment by dildick** [comment]
@slava77 So this would be the baseline implementation. The struct would have to be expanded in the future with phase-2 CSC TP parameters

**Comment by slava77** [comment]
On 11/28/17 3:09 PM, Sven Dildick wrote:
> @slava77 <https://github.com/slava77> So this would be the baseline
> implementation. The struct would have to be expanded in the future with
> phase-2 CSC TP parameters
> 

schema evolution in conddb objects is not available.
So, all expansions in this approach using structs with clearly named
variables would mean a new class to be created every time there is an
update.

**Comment by dildick** [comment]
I see. Putting the variables in a struct won't help then. So you suggested to have a single vector with values and another vector with names. Since I have `unsigned int` and `bool`, this vector should be of type `union S{unsigned int i; bool b}` right?

**Comment by davidlange6** [comment]
please test

**Comment by slava77** [comment]
On 11/28/17 3:24 PM, Sven Dildick wrote:
> I see. Putting the variables in a struct won't help then. So you
> suggested to have a single vector with values and another vector with
> names. Since I have |unsigned int| and |bool|, this vector should be of
> type |union S{unsigned int i; bool b}| right?

I don't understand why a union is needed.

**Comment by dildick** [comment]
For example, `alctNarrowMaskForR1a` is a `bool` whereas `alctEarlyTbins` is an `unsigned int`. Storing them in the same vector would require a union, not? So that it covers both types.

**Comment by dildick** [comment]
@slava77 please check the updated version. If I understand correctly, this is where you were going at.

**Comment by slava77** [comment]
On 11/28/17 4:54 PM, Sven Dildick wrote:
> For example, |alctNarrowMaskForR1a| is a |bool| whereas |alctEarlyTbins|
> is an |unsigned int|. Storing them in the same vector would require a
> union, not? So that it covers both types.
> 

IIRC my earlier suggestion, the proposal was to have one vector per type
of obects possibly needed (vector of floats, a vector of ints, and if
you with a vector of bools)

**Comment by dildick** [comment]
The proposal I have now is pretty clean. Are you saying that a union would not work? Is it absolutely required to split up the data into its different types (unsigned int, bool)?

**Comment by davidlange6** [comment]
please test

**Comment by slava77** [comment]
On 11/28/17 6:34 PM, Sven Dildick wrote:
> The proposal I have now is pretty clean. Are you saying that a union
> would not work? Is it absolutely required to split up the data into its
> different types (unsigned int, bool)?

I can't justify a union use in this case.

**Comment by dildick** [comment]
@ggovi Can you please comment?

**Comment by dildick** [comment]
@slava77 Other than having two vectors, one for each type. What other solutions are on the table?

**Comment by dildick** [comment]
@ggovi Can you point me to a twiki page that describes in detail how information is pushed to the CondDB and what constraints there are to the implementation? While @slava77 may have grounds to  object to the PR, the reason is not clear to me a this point.

**Comment by dildick** [comment]
@slava77 I split the vector up in one for `bool` and one of `int` parameters

**Comment by davidlange6** [comment]
please test

**Comment by ggovi** [comment]
+1

**Comment by dildick** [comment]
@arunhep Do you have any comments?

File: CondFormats/CSCObjects/interface/CSCDBL1TPParametersExtended.h [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+  bool getValueBool(const std::string&) const;
+
+  void setValue(const std::string&, int);
+  void setValue(const std::string&, bool);
+
+ private:
+  std::vector<int> paramsInt_;
+  std::vector<int> paramsBool_;
+
+  std::vector<std::string> paramNamesBool_{
```
Comment (ghellwig) [inline CondFormats/CSCObjects/interface/CSCDBL1TPParametersExtended.h]: @dildick purely from this class, I don't see why these string vectors are not `const`

**Review by ghellwig — COMMENTED** [review]
As @ggovi approved it already, my only comment is on the constness of the string-vector members.

**Comment by ghellwig** [comment]
please test

**Comment by ghellwig** [comment]
+1
provided jenkins agrees

**Comment by dildick** [comment]
Are there any more comments?

**Comment by dildick** [comment]
@ggovi can you please sign again?

**Comment by ggovi** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Pull request merged**